compiler/rustc_hir_typeck/src/expr_use_visitor.rs

```diff
@@ -1185,12 +1185,7 @@ impl<'tcx, Cx: TypeInformationCtxt<'tcx>, D: Delegate<'tcx>> ExprUseVisitor<'tcx
             Some(ty) => {
                 let ty = self.cx.resolve_vars_if_possible(ty);
                 self.cx.error_reported_in_ty(ty)?;
-                if ty.is_ty_var() {
-                    debug!("resolve_type_vars_or_bug: infer var from {:?}", ty);
-                    Err(self.cx.report_bug(self.cx.tcx().hir_span(id), "encountered type variable"))
-                } else {
-                    Ok(ty)
-                }
+                Ok(ty)
```

Comment:
this function is called `resolve_type_vars_or_bug`.  maybe `expect_and_resolve_type_var`?

Also, we could just emit a `ty::Error` if it fails instead of neednig a `Result`. We can still ICE if we don't expect `None`